File: genshin-builder-app/src/lib/__tests__/domain-golden.test.ts
/**
 * shared/domain-golden/cases.json を読み、Web ドメイン実装とパリティ検証する。
 */

import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import {
  calcPieceScore,
  inferScoreType,
  type ScoreType,
} from "@/lib/artifact-score";
import { makeBookmarkId } from "@/lib/bookmark-storage";
import {
  makeItemSourceKey,
  makeRangeSourceKey,
} from "@/lib/bookmark-utils";
import {
  clampInt,
  getNextStageRequirements,
  snapToLevelMark,
  type PromoteStage,
} from "@/lib/level-progression";
import {
  getRangeLevelRequirements,
  getRangeTalentRequirements,
} from "@/lib/material-requirements";
import { mainStatValue } from "@/lib/stats";
import type { TalentLevelUpgrade } from "@/lib/talent-progression";
import { snapTalentLevel } from "@/lib/talent-progression";
import { getWeaponExpBetweenMarks } from "@/lib/weapon-exp";
import type { CultivationKind } from "@/types/bookmark";

const GOLDEN_PATH = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "../../../../shared/domain-golden/cases.json",
);

interface GoldenFile {
  version: number;
  suites: Record<string, { cases: GoldenCase[] }>;
}

interface GoldenCase {
  id: string;
  fn?: string;
  input: Record<string, unknown>;
  expected: unknown;
}

const character = {
  characterId: "hu-tao",
  characterName: "胡桃",
  characterIconUrl: "https://example.com/hu-tao.png",
};

function loadGolden(): GoldenFile {
  return JSON.parse(readFileSync(GOLDEN_PATH, "utf8")) as GoldenFile;
}

function linesByMaterialId(
  lines: { materialId: string; count: number; isMora?: boolean }[],
): Record<string, { count: number; isMora: boolean }> {
  const map: Record<string, { count: number; isMora: boolean }> = {};
  for (const line of lines) {
    map[line.materialId] = {
      count: line.count,
      isMora: line.isMora === true,
    };
  }
  return map;
}

describe("domain golden parity (shared/domain-golden)", () => {
  const golden = loadGolden();

  it("loads golden file", () => {
    expect(golden.version).toBe(1);
    expect(Object.keys(golden.suites).length).toBeGreaterThan(0);
  });

  describe("clampInt", () => {
    for (const c of golden.suites.clampInt.cases) {
      it(c.id, () => {
        const { value, min, max } = c.input as {
          value: unknown;
          min: number;
          max: number;
        };
        expect(clampInt(value, min, max)).toBe(c.expected);
      });
    }
  });

  describe("snapToLevelMark", () => {
    for (const c of golden.suites.snapToLevelMark.cases) {
      it(c.id, () => {
        const { value } = c.input as { value: unknown };
        expect(snapToLevelMark(value)).toBe(c.expected);
      });
    }
  });

  describe("getWeaponExpBetweenMarks", () => {
    for (const c of golden.suites.getWeaponExpBetweenMarks.cases) {
      it(c.id, () => {
        const { from, to, rarity } = c.input as {
          from: number;
          to: number;
          rarity: number;
        };
        expect(getWeaponExpBetweenMarks(from, to, rarity)).toBe(c.expected);
      });
    }
  });

  describe("bookmarkKeys", () => {
    for (const c of golden.suites.bookmarkKeys.cases) {
      it(c.id, () => {
        const input = c.input;
        if (c.fn === "makeBookmarkId") {
          expect(
            makeBookmarkId(
              String(input.sourceKey),
              String(input.materialId),
            ),
          ).toBe(c.expected);
          return;
        }

        const ctx = {
          kind: input.kind as CultivationKind,
          targetId: String(input.targetId),
          targetName: String(input.targetName),
          subLabel: input.subLabel as string | undefined,
          character,
        };

        if (c.fn === "makeRangeSourceKey") {
          expect(
            makeRangeSourceKey(ctx, Number(input.from), Number(input.to)),
          ).toBe(c.expected);
          return;
        }

        if (c.fn === "makeItemSourceKey") {
          expect(
            makeItemSourceKey(
              ctx,
              input.scope as "next" | "stage",
              String(input.materialId),
            ),
          ).toBe(c.expected);
          return;
        }

        throw new Error(`Unknown bookmarkKeys fn: ${c.fn}`);
      });
    }
  });

  describe("getNextStageRequirements", () => {
    for (const c of golden.suites.getNextStageRequirements.cases) {
      it(c.id, () => {
        const input = c.input as {
          currentLevel: number;
          kind: "character" | "weapon";
          weaponRarity: number;
          promotes: PromoteStage[];
        };
        const stage = getNextStageRequirements(
          input.currentLevel,
          input.promotes,
          input.kind,
          input.weaponRarity,
        );
        expect(stage).not.toBeNull();
        const expected = c.expected as {
          fromLevel: number;
          toLevel: number;
          expTotal: number;
          mora: number;
          materialsById: Record<string, number>;
          levelUpMaterialIds: string[];
        };
        expect(stage!.fromLevel).toBe(expected.fromLevel);
        expect(stage!.toLevel).toBe(expected.toLevel);
        expect(stage!.expTotal).toBe(expected.expTotal);
        expect(stage!.mora).toBe(expected.mora);
        const materialsById: Record<string, number> = {};
        for (const m of stage!.materials) {
          materialsById[m.materialId] = m.count;
        }
        expect(materialsById).toEqual(expected.materialsById);
        expect(stage!.levelUpMaterials.map((m) => m.materialId)).toEqual(
          expected.levelUpMaterialIds,
        );
      });
    }
  });

  describe("getRangeLevelRequirements", () => {
    for (const c of golden.suites.getRangeLevelRequirements.cases) {
      it(c.id, () => {
        const input = c.input as {
          fromLevel: number;
          toLevel: number;
          kind: "character" | "weapon";
          promotes: PromoteStage[];
        };
        const lines = getRangeLevelRequirements(
          input.fromLevel,
          input.toLevel,
          input.promotes,
          input.kind,
        );
        const expected = c.expected as {
          linesByMaterialId: Record<
            string,
            { count: number; isMora?: boolean }
          >;
        };
        expect(linesByMaterialId(lines)).toEqual(expected.linesByMaterialId);
      });
    }
  });

  describe("getRangeTalentRequirements", () => {
    for (const c of golden.suites.getRangeTalentRequirements.cases) {
      it(c.id, () => {
        const input = c.input as {
          fromLevel: number;
          toLevel: number;
          maxLevel: number;
          upgrades: TalentLevelUpgrade[];
        };
        const lines = getRangeTalentRequirements(
          input.fromLevel,
          input.toLevel,
          input.maxLevel,
          input.upgrades,
        );
        const expected = c.expected as {
          linesByMaterialId: Record<string, { count: number }>;
        };
        const actual: Record<string, { count: number }> = {};
        for (const line of lines) {
          actual[line.materialId] = { count: line.count };
        }
        expect(actual).toEqual(expected.linesByMaterialId);
      });
    }
  });

  describe("snapTalentLevel", () => {
    for (const c of golden.suites.snapTalentLevel.cases) {
      it(c.id, () => {
        const { value, max } = c.input as { value: unknown; max: number };
        expect(snapTalentLevel(value, max)).toBe(c.expected);
      });
    }
  });

  describe("artifactMainStatValue", () => {
    for (const c of golden.suites.artifactMainStatValue.cases) {
      it(c.id, () => {
        const { statName, level } = c.input as {
          statName: string;
          level: number;
        };
        expect(mainStatValue(statName, level)).toBe(c.expected);
      });
    }
  });

  describe("inferScoreType", () => {
    for (const c of golden.suites.inferScoreType.cases) {
      it(c.id, () => {
        const { specialProp, name } = c.input as {
          specialProp: string | null;
          name: string;
        };
        expect(inferScoreType(specialProp, name)).toBe(c.expected);
      });
    }
  });

  describe("calcPieceScore", () => {
    for (const c of golden.suites.calcPieceScore.cases) {
      it(c.id, () => {
        const input = c.input as {
          type: ScoreType;
          substats: { stat: string; value: number }[];
        };
        const piece = {
          setId: "",
          level: 0,
          mainStat: "",
          substats: input.substats,
        };
        expect(calcPieceScore(piece, input.type)).toBe(c.expected);
      });
    }
  });
});

File: genshin-builder-mobile/test/domain/domain_golden_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:genshin_builder_mobile/domain/artifact_score.dart';
import 'package:genshin_builder_mobile/domain/bookmark_utils.dart';
import 'package:genshin_builder_mobile/domain/character_stats.dart';
import 'package:genshin_builder_mobile/domain/level_progression.dart';
import 'package:genshin_builder_mobile/domain/material_requirements.dart';
import 'package:genshin_builder_mobile/domain/models/artifact_state.dart';
import 'package:genshin_builder_mobile/domain/models/bookmark.dart';
import 'package:genshin_builder_mobile/domain/models/calculation_models.dart';
import 'package:genshin_builder_mobile/domain/talent_progression.dart';
import 'package:genshin_builder_mobile/domain/weapon_exp.dart';

/// shared/domain-golden/cases.json を読み、Dart ドメイン実装とパリティ検証する。
void main() {
  final golden = _loadGolden();

  test('loads golden file', () {
    expect(golden['version'], 1);
    expect((golden['suites'] as Map).isNotEmpty, isTrue);
  });

  final suites = golden['suites'] as Map<String, dynamic>;

  group('clampInt', () {
    for (final c in _cases(suites, 'clampInt')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        expect(
          clampInt(input['value'], input['min'] as int, input['max'] as int),
          c['expected'],
        );
      });
    }
  });

  group('snapToLevelMark', () {
    for (final c in _cases(suites, 'snapToLevelMark')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        expect(snapToLevelMark(input['value']), c['expected']);
      });
    }
  });

  group('getWeaponExpBetweenMarks', () {
    for (final c in _cases(suites, 'getWeaponExpBetweenMarks')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        expect(
          getWeaponExpBetweenMarks(
            input['from'] as int,
            input['to'] as int,
            input['rarity'] as int,
          ),
          c['expected'],
        );
      });
    }
  });

  group('bookmarkKeys', () {
    const character = BookmarkCharacterSource(
      characterId: 'hu-tao',
      characterName: '胡桃',
      characterIconUrl: 'https://example.com/hu-tao.png',
    );

    for (final c in _cases(suites, 'bookmarkKeys')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        final fn = c['fn'] as String?;

        if (fn == 'makeBookmarkId') {
          expect(
            makeBookmarkId(
              input['sourceKey'] as String,
              input['materialId'] as String,
            ),
            c['expected'],
          );
          return;
        }

        final ctx = CultivationBookmarkContext(
          kind: _parseKind(input['kind'] as String),
          targetId: input['targetId'] as String,
          targetName: input['targetName'] as String,
          subLabel: input['subLabel'] as String?,
          character: character,
        );

        if (fn == 'makeRangeSourceKey') {
          expect(
            makeRangeSourceKey(ctx, input['from'] as int, input['to'] as int),
            c['expected'],
          );
          return;
        }

        if (fn == 'makeItemSourceKey') {
          expect(
            makeItemSourceKey(
              ctx,
              input['scope'] as String,
              input['materialId'] as String,
            ),
            c['expected'],
          );
          return;
        }

        fail('Unknown bookmarkKeys fn: $fn');
      });
    }
  });

  group('getNextStageRequirements', () {
    for (final c in _cases(suites, 'getNextStageRequirements')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        final promotes =
            (input['promotes'] as List)
                .map((p) => _parsePromote(p as Map<String, dynamic>))
                .toList();
        final stage = getNextStageRequirements(
          input['currentLevel'] as int,
          promotes,
          input['kind'] as String,
          input['weaponRarity'] as int,
        );
        expect(stage, isNotNull);
        final expected = c['expected'] as Map<String, dynamic>;
        expect(stage!.fromLevel, expected['fromLevel']);
        expect(stage.toLevel, expected['toLevel']);
        expect(stage.expTotal, expected['expTotal']);
        expect(stage.mora, expected['mora']);

        final materialsById = <String, int>{
          for (final m in stage.materials) m.materialId: m.count,
        };
        expect(
          materialsById,
          Map<String, int>.from(expected['materialsById'] as Map),
        );
        expect(
          stage.levelUpMaterials.map((m) => m.materialId).toList(),
          List<String>.from(expected['levelUpMaterialIds'] as List),
        );
      });
    }
  });

  group('getRangeLevelRequirements', () {
    for (final c in _cases(suites, 'getRangeLevelRequirements')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        final promotes =
            (input['promotes'] as List)
                .map((p) => _parsePromote(p as Map<String, dynamic>))
                .toList();
        final lines = getRangeLevelRequirements(
          input['fromLevel'] as int,
          input['toLevel'] as int,
          promotes,
          input['kind'] as String,
        );
        final expected = c['expected'] as Map<String, dynamic>;
        final expectedMap =
            expected['linesByMaterialId'] as Map<String, dynamic>;
        final actual = <String, Map<String, Object?>>{};
        for (final line in lines) {
          actual[line.materialId] = {
            'count': line.count,
            if (line.isMora) 'isMora': true else 'isMora': false,
          };
        }
        final normalizedExpected = <String, Map<String, Object?>>{};
        for (final entry in expectedMap.entries) {
          final v = Map<String, Object?>.from(entry.value as Map);
          normalizedExpected[entry.key] = {
            'count': v['count'],
            'isMora': v['isMora'] ?? false,
          };
        }
        expect(actual, normalizedExpected);
      });
    }
  });

  group('getRangeTalentRequirements', () {
    for (final c in _cases(suites, 'getRangeTalentRequirements')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        final upgrades =
            (input['upgrades'] as List)
                .map((u) => _parseTalentUpgrade(u as Map<String, dynamic>))
                .toList();
        final lines = getRangeTalentRequirements(
          input['fromLevel'] as int,
          input['toLevel'] as int,
          input['maxLevel'] as int,
          upgrades,
        );
        final expected = c['expected'] as Map<String, dynamic>;
        final expectedMap =
            expected['linesByMaterialId'] as Map<String, dynamic>;
        final actual = <String, Map<String, Object?>>{};
        for (final line in lines) {
          actual[line.materialId] = {'count': line.count};
        }
        final normalizedExpected = <String, Map<String, Object?>>{};
        for (final entry in expectedMap.entries) {
          final v = Map<String, Object?>.from(entry.value as Map);
          normalizedExpected[entry.key] = {'count': v['count']};
        }
        expect(actual, normalizedExpected);
      });
    }
  });

  group('snapTalentLevel', () {
    for (final c in _cases(suites, 'snapTalentLevel')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        expect(
          snapTalentLevel(input['value'], input['max'] as int),
          c['expected'],
        );
      });
    }
  });

  group('artifactMainStatValue', () {
    for (final c in _cases(suites, 'artifactMainStatValue')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        expect(
          artifactMainStatValue(
            input['statName'] as String,
            input['level'] as int,
          ),
          c['expected'],
        );
      });
    }
  });

  group('inferScoreType', () {
    for (final c in _cases(suites, 'inferScoreType')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        expect(
          artifactScoreTypeToStorage(
            inferScoreType(
              input['specialProp'] as String?,
              input['name'] as String,
            ),
          ),
          c['expected'],
        );
      });
    }
  });

  group('calcPieceScore', () {
    for (final c in _cases(suites, 'calcPieceScore')) {
      test(c['id'] as String, () {
        final input = c['input'] as Map<String, dynamic>;
        final type = artifactScoreTypeFromString(input['type'] as String)!;
        final piece = ArtifactPiece(
          substats: [
            for (final raw in input['substats'] as List)
              ArtifactSubstat(
                stat: (raw as Map)['stat'] as String,
                value: (raw['value'] as num).toDouble(),
              ),
          ],
        );
        expect(calcArtifactPieceScore(piece, type), c['expected']);
      });
    }
  });
}

Map<String, dynamic> _loadGolden() {
  final candidates = [
    File('${Directory.current.path}/../shared/domain-golden/cases.json'),
    File('${Directory.current.path}/shared/domain-golden/cases.json'),
  ];
  for (final file in candidates) {
    if (file.existsSync()) {
      return jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
    }
  }
  fail(
    'shared/domain-golden/cases.json not found. '
    'Run flutter test from genshin-builder-mobile/.',
  );
}

List<Map<String, dynamic>> _cases(Map<String, dynamic> suites, String name) {
  final suite = suites[name] as Map<String, dynamic>;
  return (suite['cases'] as List)
      .map((e) => Map<String, dynamic>.from(e as Map))
      .toList();
}

CultivationKind _parseKind(String kind) => switch (kind) {
  'character-level' => CultivationKind.characterLevel,
  'weapon-level' => CultivationKind.weaponLevel,
  'talent' => CultivationKind.talent,
  _ => throw ArgumentError('Unknown kind: $kind'),
};

PromoteStage _parsePromote(Map<String, dynamic> json) {
  final costItems = <String, int>{
    for (final e in (json['costItems'] as Map).entries)
      e.key as String: (e.value as num).toInt(),
  };
  return PromoteStage(
    promoteLevel: json['promoteLevel'] as int,
    unlockMaxLevel: json['unlockMaxLevel'] as int,
    costItems: costItems,
    coinCost: json['coinCost'] as int,
  );
}

TalentLevelUpgrade _parseTalentUpgrade(Map<String, dynamic> json) {
  final costItems = <String, int>{
    for (final e in (json['costItems'] as Map).entries)
      e.key as String: (e.value as num).toInt(),
  };
  return TalentLevelUpgrade(
    level: json['level'] as int,
    costItems: costItems,
    coinCost: json['coinCost'] as int,
  );
}
